react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx

```diff
@@ -0,0 +1,172 @@
+import { useMemo, useCallback, useEffect, useReducer } from 'react'
+
+const useTableTreeCheckboxes = ({
+  items,
+  columns,
+  onToggle,
+}: hookInput): checkboxesHookReturn => {
+  const [checkedItems, dispatch] = useReducer(reducer, [])
+
+  const itemTree = useMemo(() => {
+    return { id: 'root', children: items.map((item, i) => parseTree(item, i)) }
+  }, [items, columns])
+
+  const toggle = useCallback(
+    (item: ItemTree): void => {
+      dispatch({ type: ActionType.Toggle, item })
+    },
+    [checkedItems]
+  )
+
+  useEffect(() => {
+    onToggle({ checkedItems })
+  }, [toggle])
+
+  useEffect(() => {
+    const shake = (tree: ItemTree) => {
+      const { children } = tree
+      const checkedIds = idsFrom(checkedItems)
+
+      if (!children) return
+
+      const areChildsChecked = children.every(child =>
+        checkedIds.includes(child.id)
+      )
+      const isRootChecked = checkedIds.includes(tree.id)
+
+      if (areChildsChecked && !isRootChecked)
+        dispatch({ type: ActionType.Check, item: tree })
+
+      if (!areChildsChecked && isRootChecked)
+        dispatch({ type: ActionType.Uncheck, item: tree })
+
+      children.forEach(shake)
+    }
+    shake(itemTree)
+  }, [checkedItems])
+
+  const isChecked = useCallback(
+    (item: ItemTree) => {
+      const { children } = item
+      const condition = children
+        ? children.every(child => idsFrom(checkedItems).includes(child.id))
+        : checkedItems.some(row => row.id === item.id)
+      return condition
+    },
+    [checkedItems]
+  )
+
+  const isPartiallyChecked = useCallback(
+    (item: ItemTree) => {
+      const { children } = item
+      const flatChildren = getFlat(item).slice(1)
+      return (
+        children &&
+        flatChildren.some(child => idsFrom(checkedItems).includes(child.id))
+      )
+    },
+    [checkedItems]
+  )
+
+  return { checkedItems, isChecked, isPartiallyChecked, itemTree, toggle }
+}
+
+function reducer(state: Array<ItemTree>, action: Action) {
+  switch (action.type) {
+    case ActionType.Check: {
+      return [...state, action.item]
+    }
+    case ActionType.Uncheck: {
+      return state.filter(row => row.id !== action.item.id)
+    }
+    case ActionType.Toggle: {
+      const { item } = action
+
+      if (!item) return state
+
+      if (!item.children) {
+        return state.some(row => row.id === item.id)
+          ? state.filter(row => row.id !== item.id)
+          : [...state, item]
+      } else {
+        const flatItems = getFlat(item)
+        return state.some(row => row.id === item.id)
+          ? state.filter(row => !idsFrom(flatItems).includes(row.id))
+          : [...state, ...flatItems].reduce(
+              (acc, item) =>
+                idsFrom(acc).includes(item.id) ? acc : [...acc, item],
+              []
```

Comment:
:developers:

Reply:
Made some changes, so that code is more readable. 😅